### PR TITLE
Newton basis

### PR DESCRIPTION
**Pull request by viktorya-grigoryan**
## Linked Issue
Fixes # (issue number)

## Implementation Checklist
- [x] My class inherits from `PolynomialBasis` (if applicable).
- [x] I have implemented all abstract methods (`evaluate_basis`, `fit`, etc.).
- [x] I have used `numpy` vectorization to avoid inefficient Python loops.
- [x] Documentation strings (docstrings) are provided for all public methods.

### 1. Mathematical Integrity
- [ ] **Convergence:** Does the implementation maintain expected accuracy? (Checked for Runge's phenomenon or stability).
- [ ] **Edge Cases:** Does it handle single points or non-equally spaced nodes correctly?
- [ ] **Domain Mapping:** Does the class correctly map the input domain to the basis domain (e.g., [-1, 1] for Chebyshev)?

### 2. API & Style
- [ ] **Method Signatures:** Methods match the Abstract Base Class exactly.
- [ ] **Naming:** Follows `CamelCase` for classes and `snake_case` for functions/variables.
- [ ] **Clean Code:** No "magic numbers"; constants are named and explained.

### 3. Testing Quality
- [ ] **Coverage:** Unit tests cover typical usage and edge cases.
- [ ] **Precision:** Tests use `np.testing.assert_allclose` instead of `==`.
- [ ] **Validation:** There is a test proving the basis can exactly reconstruct a polynomial of its own degree.

## Screenshots / Plots

### CODE REVIEW

File: src/numanalysislib/basis/newton.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            return np.ones_like(x, dtype=float)
+
+        result = np.ones_like(x, dtype=float)
+        for j in range(index):
+            result *= (x - self.x_nodes[j])
+
+        return result
+
+ 
+    # 2. Divided Differences (fit)
```
Comment (mtezzele) [inline src/numanalysislib/basis/newton.py]: please remove these type of comments and properly document each method

File: tests/test_newton.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+    x_nodes = np.array([0.0, 1.0])  # wrong size
+    y_nodes = np.array([0.0, 1.0])
+
+    with pytest.raises(ValueError):
+        basis.fit(x_nodes, y_nodes)
+
+
+
+# 7. Test exact match
```
Comment (mtezzele) [inline tests/test_newton.py]: remove also this type of comments and if needed put them inside the methods. Also please remove many of the unnecessary empty lines

File: src/numanalysislib/basis/newton.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+        super().__init__(degree=degree,
+                         a=np.min(x_nodes),
+                         b=np.max(x_nodes))
+
+
+    """
+    Basis functions
+    n_i(x) = Π_{j=0}^{i-1} (x - x_j)
+    """
```
Comment (xc900606) [inline src/numanalysislib/basis/newton.py]: It would be clearer to move that formula into evaluate_basis’s docstring and document parameters/returns there.

File: src/numanalysislib/basis/newton.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+        coef = y_nodes.astype(float).copy()
+        n = len(coef)
+
+        # In-place divided differences
+        for j in range(1, n):
+            coef[j:n] = (
+                (coef[j:n] - coef[j-1:n-1]) /
+                (x_nodes[j:n] - x_nodes[0:n-j])
+            )
```
Comment (xc900606) [inline src/numanalysislib/basis/newton.py:85]: Would it make sense to add an edge-case test for repeated nodes? Repeated nodes can lead to zero denominators, so it would be good to define whether this should raise a `ValueError`.

File: src/numanalysislib/basis/newton.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+        if index == 0:
+            return np.ones_like(x, dtype=float)
+
+        result = np.ones_like(x, dtype=float)
+        for j in range(index):
+            result *= (x - self.x_nodes[j])
+
+        return result
+
```
Comment (xc900606) [inline src/numanalysislib/basis/newton.py:50]: Should we add a check for a negative index here? Right now, if someone passes -1, it seems to return 1s rather than raise an error.

File: src/numanalysislib/basis/newton.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        Returns
+        -------
+        np.ndarray
+            Array of Newton coefficients [c_0, ..., c_{n-1}].
+        """
+        x_nodes = np.asarray(x_nodes)
+        y_nodes = np.asarray(y_nodes)
+
+        if len(x_nodes) != self.n_dofs:
+            raise ValueError("Number of nodes must match basis dimension")
```
Comment (xc900606) [inline src/numanalysislib/basis/newton.py:69]: Should we also check that y_nodes has the expected length here? since each x_node should have a matching value

**Review by xc900606 — COMMENTED** [review]
Overall, the Newton basis implementation looks good. I left a few comments, mainly on edge cases, especially around invalid index handling, repeated nodes in the divided-difference step. The core functionality and main tests look good overall.

**Comment by viktorya-grigoryan** [comment]
Fixed all the comments
